dm-agent/src/main.rs

```diff
@@ -0,0 +1,190 @@
+use futures::future::FutureExt;
+use mqtt_client::{Client, Message, MessageStream, QoS, Topic, TopicFilter};
+use std::str::from_utf8;
+use std::sync::Arc;
+use thin_edge_json::{
+    group::MeasurementGrouper, measurement::current_timestamp, measurement::FlatMeasurementVisitor,
+    serialize::ThinEdgeJsonSerializer,
+};
+use tokio::select;
+use tokio::time::Duration;
+use tracing::{debug_span, error, info, instrument, Instrument};
+
+const DEFAULT_STATS_COLLECTION_INTERVAL: u64 = 1000;
+
+const DEFAULT_HOST: &str = "localhost";
+const DEFAULT_PORT: u16 = 1883;
+const CLIENT_ID: &str = "tedge-dm-agent";
+const SOURCE_TOPIC: &str = "collectd/#";
+const TARGET_TOPIC: &str = "tedge/measurements";
+
+const APP_NAME: &str = "tedge-dm-agent";
+const DEFAULT_LOG_LEVEL: &str = "warn";
+const TIME_FORMAT: &str = "%Y-%m-%dT%H:%M:%S%.3f%:z";
+
+#[tokio::main]
+async fn main() {
+    let filter = std::env::var("RUST_LOG").unwrap_or_else(|_| DEFAULT_LOG_LEVEL.into());
+    tracing_subscriber::fmt()
+        .with_timer(tracing_subscriber::fmt::time::ChronoUtc::with_format(
+            TIME_FORMAT.into(),
+        ))
+        .with_env_filter(filter)
+        .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
+        .init();
+
+    info!("{} starting!", APP_NAME);
+
+    if let Err(err) = run().instrument(debug_span!(APP_NAME)).await {
+        error!("{}", err);
+    }
+}
+
+#[instrument(name = "main")]
+async fn run() -> Result<(), anyhow::Error> {
+    let config = mqtt_client::Config::new(DEFAULT_HOST, DEFAULT_PORT);
+    let mqtt = Arc::new(Client::connect(CLIENT_ID, &config).await?);
+    let filter = TopicFilter::new(SOURCE_TOPIC)?.qos(QoS::AtMostOnce);
+
+    let mut errors = mqtt.subscribe_errors();
+    let mut messages: MessageStream = mqtt.subscribe(filter).await?;
+
+    let mut interval =
```

Comment:
Is that correct here? It feels that the statement to start the 1s interval should be started when the first message is received (it was specified in the doc like that).
However, does this code just raises "interval.tick()" every 1s, starting at a more-or-less ramdom time?

Reply:
@vaillant your observation is correct. this is mainly to keep the code as simple as possible (for now).

Reply:
OK, but then this code is not correct and should be changed.